pyfr/readers/cgns.py

```diff
@@ -0,0 +1,435 @@
+
+            name = 'fluid'
+            for bcname, bcrng in bc.items():
+                if elerng[0] >= bcrng[0] and elerng[1] <= bcrng[1]:
+                    name = bcname
+
+            pent = pents.setdefault(name)
+            if pent is None:
+                newp += 1
+                pents[name] = pent = newp
+
+            elenodes.update({(k, pent): v for k, v in elenode.items()})
+
+    def _read_nodepts(self, zone):
+        nnode = zone['size'][0]
+        nodepts = np.zeros((3, nnode))
+
+        self._cgns.coord_read(zone, 'CoordinateX', nodepts[0])
+        self._cgns.coord_read(zone, 'CoordinateY', nodepts[1])
+        self._cgns.coord_read(zone, 'CoordinateZ', nodepts[2])
+
+        return nodepts
+
+    def _read_bc(self, zone):
+        nbc = self._cgns.nbocos(zone)
+        bc = {}
+
+        for idx_bc in range(nbc):
+            boco = self._cgns.boco_read(zone, idx_bc)
+            name = boco['name'].lower()
+            bc[name] = boco['range']
+
+        return bc
+
+    def _read_element(self, zone, idx):
+        s = self._cgns.section_read(zone, idx)
+
+        elerng = s['range']
+        conn = np.zeros(s['dim'], dtype=np.int32)
+        self._cgns.elements_read(s, conn)
+
+        cgns_type = s['etype']
+        petype = self._cgns_map[cgns_type][0]
+
+        elenode = {}
+
+        if petype == 'mixed':
+            i = 0
+            mele = defaultdict(list)
+
+            while i < s['dim']:
+                try:
+                    cgns_type = conn[i]
+                    petype, spts = self._cgns_map[cgns_type]
+                except KeyError as err:
+                    raise KeyError('Unsupported elements: {}'.format(err))
+
+                mele[cgns_type].append(conn[i + 1: i + 1 + spts])
+                i += 1 + spts
+
+            elenode = {k: np.array(v) for k, v in mele.items()}
+        else:
+            spts = self._cgns_map[cgns_type][1]
+            elenode[cgns_type] = conn.reshape(-1, spts)
+
+        return elerng, elenode
+
+
+class CGNSReader(BaseReader):
+    # Supported file types and extensions
+    name = 'cgns'
+    extn = ['.cgns']
+
+    # CGNS element types to PyFR type (petype) and node counts
+    etype_map = CGNSZoneReader._cgns_map
+
+    # First-order node numbers associated with each element face
+    petype_fnmap = {
+        'tri': {'line': [[0, 1], [1, 2], [2, 0]]},
+        'quad': {'line': [[0, 1], [1, 2], [2, 3], [3, 0]]},
+        'tet': {'tri': [[0, 1, 2], [0, 1, 3], [0, 2, 3], [1, 2, 3]]},
+        'hex': {'quad': [[0, 1, 2, 3], [0, 1, 4, 5], [1, 2, 5, 6],
+                         [2, 3, 6, 7], [0, 3, 4, 7], [4, 5, 6, 7]]},
+        'pri': {'quad': [[0, 1, 3, 4], [1, 2, 4, 5], [0, 2, 3, 5]],
+                'tri': [[0, 1, 2], [3, 4, 5]]},
+        'pyr': {'quad': [[0, 1, 2, 3]],
+                'tri': [[0, 1, 4], [1, 2, 4], [2, 3, 4], [0, 3, 4]]}
+    }
+
+    # Nodemap to PyFR node order
+    nodemaps_from_pyfr = CGNSNodeMaps.from_pyfr
```

Comment:
Might be better if we do `nodemaps = CGNSNodeMaps` (and a similar thing in the Gmsh reader).